WDE.SmartScriptEditor/History/SmartDataListHistoryHandler.cs

```diff
@@ -0,0 +1,139 @@
+
+        public SmartDataListHistoryHandler(ObservableCollection<SmartGenericJsonData> smartDataList)
+        {
+            this.smartDataList = smartDataList;
+            smartDataList.CollectionChanged += OnDataCollectionChanged;
+        }
+
+        public void Dispose()
+        {
+            smartDataList.CollectionChanged -= OnDataCollectionChanged;
+        }
+
+        private void OnDataCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+        {
+            if (e.Action == NotifyCollectionChangedAction.Add)
+            {
+                if (e.NewItems != null)
+                {
+                    foreach (SmartGenericJsonData data in e.NewItems)
+                        PushAction(new SmartDataListHistoryAction(in data, e.NewStartingIndex, ListHistoryActionMode.ITEM_ADDED, smartDataList));
+                }
+            }
+            else if (e.Action == NotifyCollectionChangedAction.Remove)
+            {
+                if (e.OldItems != null)
+                {
+                    foreach(SmartGenericJsonData data in e.OldItems)
+                        PushAction(new SmartDataListHistoryAction(in data, e.OldStartingIndex, ListHistoryActionMode.ITEM_REMOVED, smartDataList));
+                }
+            }
+            else if (e.Action == NotifyCollectionChangedAction.Replace)
+            {
+                if (e.NewItems != null && e.OldItems != null && e.NewItems.Count == e.OldItems.Count)
+                {
+                    for (int i = 0; i < e.NewItems.Count; ++i)
+                    {
+                        SmartGenericJsonData oldItem = e.OldItems[i] is SmartGenericJsonData ? (SmartGenericJsonData) e.OldItems[i] : default;
+                        SmartGenericJsonData newItem = e.NewItems[i] is SmartGenericJsonData ? (SmartGenericJsonData) e.NewItems[i] : default;
+                        PushAction(new SmartDataListHistoryReplaceAction(in oldItem, in newItem, e.NewStartingIndex, smartDataList));
+                    }
+                }
+            }
+        }
+    }
+
+    internal enum ListHistoryActionMode
+    {
+        ITEM_ADDED,
+        ITEM_REMOVED,
+    }
+    
+    internal class SmartDataListHistoryAction : IHistoryAction
+    {
+        private readonly SmartGenericJsonData element;
+        private readonly int index;
+        private readonly ListHistoryActionMode actionMode;
+        private readonly ObservableCollection<SmartGenericJsonData> collection;
+
+        internal SmartDataListHistoryAction(in SmartGenericJsonData element, int index, ListHistoryActionMode actionMode, ObservableCollection<SmartGenericJsonData> collection)
+        {
+            this.element = element;
+            this.index = index;
+            this.actionMode = actionMode;
+            this.collection = collection;
+        } 
+        
+        public void Undo()
+        {
+            if (actionMode == ListHistoryActionMode.ITEM_ADDED)
```

Comment:
Instead of having enum Add/Remove and series of `ifs` isn’t it more object oriented to just create SmartDataAddListHistoryAction and SmartDataRemoveListHistoryAction? ;P